pkg/controller/elasticsearch/nodespec/desired_nodes.go

```diff
@@ -0,0 +1,313 @@
+type ResourceNotAvailable struct {
+	nodeSet string
+	reasons []string
+}
+
+func (r ResourceNotAvailable) Error() string {
+	return fmt.Sprintf("cannot compute resources for NodeSet \"%s\": %s", r.nodeSet, strings.Join(r.reasons, ", "))
+}
+
+type nodeResources []nodeResource
+
+func (n nodeResources) requeue() bool {
+	for _, nodeResource := range n {
+		if nodeResource.requeue {
+			return true
+		}
+	}
+	return false
+}
+
+type nodeResource struct {
+	nodeName        string
+	memory, storage int64
+	cpu             client.ProcessorsRange
+	requeue         bool
+}
+
+type nodeSetResourcesBuilder struct {
+	nodeSet string
+	cpu     client.ProcessorsRange
+	memory  int64
+	reasons []string
+}
+
+func (n nodeSetResourcesBuilder) addReason(reason string) nodeSetResourcesBuilder {
+	n.reasons = append(n.reasons, reason)
+	return n
+}
+
+func (n nodeSetResourcesBuilder) toError() error {
+	if len(n.reasons) == 0 {
+		return nil
+	}
+	return &ResourceNotAvailable{nodeSet: n.nodeSet, reasons: n.reasons}
+}
+
+const (
+	envPodName             = "${" + settings.EnvPodName + "}"
+	envNamespace           = "${" + settings.EnvNamespace + "}"
+	envHeadlessServiceName = "${" + settings.HeadlessServiceName + "}"
+)
+
+// ToDesiredNodes returns the desired nodes, as expected by the desired nodes API, from an expected resources list.
+// A boolean is also returned to indicate whether a requeue should be done to set a more accurate state of the
+// storage capacity.
+func (l ResourcesList) ToDesiredNodes(
+	ctx context.Context,
+	k8sClient k8s.Client,
+	version string,
+) (desiredNodes []client.DesiredNode, requeue bool, err error) {
+	span, ctx := apm.StartSpan(ctx, "compute_desired_nodes", tracing.SpanTypeApp)
+	defer span.End()
+	desiredNodes = make([]client.DesiredNode, 0, l.ExpectedNodeCount())
+	for _, resources := range l {
+		sts := resources.StatefulSet
+		esContainer := getElasticsearchContainer(sts.Spec.Template.Spec.Containers)
+		if esContainer == nil {
+			return nil, false, fmt.Errorf("cannot find Elasticsearch container in StatefulSet %s/%s", sts.Namespace, sts.Name)
+		}
+
+		nodeResources, err := nodeSetResourcesBuilder{nodeSet: resources.NodeSet}.
+			withProcessors(esContainer.Resources).
+			withMemory(esContainer.Resources).
+			withStorage(ctx, k8sClient, sts, resources.Config, esContainer)
+		if err != nil {
+			return nil, false, err
+		}
+
+		requeue = requeue || nodeResources.requeue()
+
+		for _, nodeResource := range nodeResources {
+			// Replace variable in the configuration
+			knownVariablesReplacer := strings.NewReplacer(
+				envPodName, nodeResource.nodeName,
+				envNamespace, sts.Namespace,
+				envHeadlessServiceName, resources.HeadlessService.Name,
+			)
+			var settings map[string]interface{}
+			if err := resources.Config.CanonicalConfig.Unpack(&settings); err != nil {
+				return nil, false, err
+			}
+			visit(nil, settings, func(s string) string {
+				return knownVariablesReplacer.Replace(s)
```

Comment:
I would like to better understand why we have to replace these variables for the desired nodes API. https://github.com/elastic/cloud-on-k8s/issues/5528#issuecomment-1118508705 it seems that you are worried some of the optimisations that ES does will be negatively affected if there are still variables in the settings. Can you maybe add a comment?

Reply:
I added a comment, the main motivation is to provide a value to the `node.name` parameter.